src/databricks/labs/ucx/azure/resources.py

```diff
@@ -117,44 +121,134 @@ def inner() -> dict[str, str]:
+
+    @property
+    def token(self):
+        return self._token_source.token()
+
+
+class AzureResources:
+    def __init__(self, *, include_subscriptions=None, api_client: AzureAPIClient):
+        if not include_subscriptions:
+            include_subscriptions = []
+        self._api_client = api_client
+        self._include_subscriptions = include_subscriptions
+        self._role_definitions = {}  # type: dict[str, str]
+        self._principals: dict[str, Principal | None] = {}
+
     def _get_subscriptions(self) -> Iterable[AzureSubscription]:
-        for subscription in self._get_resource("/subscriptions", api_version="2022-12-01").get("value", []):
+        for subscription in self._api_client.get("/subscriptions", "azure_mgmt", "2022-12-01").get("value", []):
             yield AzureSubscription(
                 name=subscription["displayName"],
                 subscription_id=subscription["subscriptionId"],
                 tenant_id=subscription["tenantId"],
             )
 
-    def _tenant_id(self):
-        token = self._token_source.token()
+    def create_service_principal(self) -> Principal:
+        try:
+            application_info: dict[str, str] = self._api_client.put(
+                "/v1.0/applications", "azure_graph", {"displayName": "UCXServicePrincipal"}
+            )
+            app_id = application_info.get("appId")
+            assert app_id is not None
+            service_principal_info: dict[str, str] = self._api_client.put(
+                "/v1.0/servicePrincipals", "azure_graph", {"appId": app_id}
+            )
+            client_id = service_principal_info.get("appId")
+            assert client_id is not None
+            secret_info: dict[str, str] = self._api_client.put(
+                f"/v1.0/servicePrincipals(appId='{client_id}')/addPassword", "azure_graph"
+            )
+
+        except PermissionDenied:
+            msg = (
+                "Permission denied. Please run this cmd under the identity of a user who has"
+                " create service principal permission."
+            )
+            logger.error(msg)
+            raise PermissionDenied(msg) from None
+        secret = secret_info.get("secretText")
+        display_name = service_principal_info.get("displayName")
+        object_id = service_principal_info.get("id")
+        directory_id = service_principal_info.get("appOwnerOrganizationId")
+        assert client_id is not None
+        assert display_name is not None
+        assert object_id is not None
+        assert directory_id is not None
+        assert secret is not None
+        return Principal(client_id, display_name, object_id, directory_id, secret)
+
+    def apply_storage_permission(self, principal_id: str, resource: AzureResource, role: str):
+        try:
+            role_name = _ROLES[role]["name"]
+            role_id = _ROLES[role]["id"]
+            path = f"{str(resource)}/providers/Microsoft.Authorization/roleAssignments/{role_name}"
```

Comment:
```suggestion
            path = f"{str(resource)}/providers/Microsoft.Authorization/roleAssignments"
```
I think you need to do "create" instead of "overwrite"